core/shared/src/main/scala/org/typelevel/idna4s/core/bootstring/Damp.scala

```diff
@@ -21,44 +21,102 @@
 
 package org.typelevel.idna4s.core.bootstring
 
-import cats._
+import cats.Show
+import cats.kernel._
 import cats.syntax.all._
 
-sealed abstract class Damp extends Product with Serializable {
-  def value: Int
-
+/**
+ * Newtype for RFC 3492 Damp value.
+ *
+ * A damp is a special value used in the bias adaptation function of Bootstring when scaling the
+ * very first delta. It compensates for the fact that the second delta is usually much smaller
+ * the first.
+ *
+ * @see
+ *   [[https://datatracker.ietf.org/doc/html/rfc3492#section-3.4 Bias Adaptation]]
+ */
+final class Damp private (val value: Int) extends AnyVal {
   final override def toString: String = s"Damp(value = ${value})"
 }
 
 object Damp {
-  final private[this] case class DampImpl(override val value: Int) extends Damp
+  private def apply(value: Int): Damp =
+    new Damp(value)
 
-  val PunycodeDamp: Damp = unsafeFromInt(700)
+  val MinValue: Damp = Damp(2)
+  val MaxValue: Damp = Damp(Int.MaxValue)
 
-  def fromInt(value: Int): Either[String, Damp] =
+  /**
+   * The damp used by the Punycode Bootstring invocation.
+   */
+  final val PunycodeDamp: Damp = new Damp(700)
+
+  /**
+   * Attempt to create a [[Damp]] value from an int value, throwing an exception if it is
+   * outside the valid domain for a [[Damp]] value.
+   */
+  def unsafeFromInt(value: Int): Damp =
     if (value >= 2) {
-      Right(DampImpl(value))
+      Damp(value)
     } else {
-      Left(s"According to RFC-3492 damp values must be >= 2.")
+      throw new IllegalArgumentException(
+        s"According to RFC-3492 damp values must be >= 2: ${value}")
     }
 
-  def unsafeFromInt(value: Int): Damp =
-    fromInt(value).fold(
-      e => throw new IllegalArgumentException(e),
-      identity
-    )
+  /**
+   * Attempt to create a [[Damp]] value from a `String` value, representing a base 10 int32
+   * value, throwing an exception if the `String` is not a valid base 10 int32 value or if it is
+   * outside the valid domain for a [[Damp]] value.
+   */
+  def unsafeFromString(value: String): Damp =
+    unsafeFromInt(value.toInt)
 
+  /**
+   * Attempt to create a [[Damp]] value from an int value, yielding an error if it is outside
+   * the valid domain for a [[Damp]] value.
+   */
+  def fromInt(value: Int): Either[String, Damp] =
+    Either.catchNonFatal(unsafeFromInt(value)).leftMap(_.getLocalizedMessage)
```

Comment:
Another one of my micro-optimizations: try/catch can be expensive and bad for machine optimization (particularly on JS and Native platforms I think). So I wonder if it may be better to express the unsafe variant in terms of the safe variant

I wouldn't worry about it in this PR, but maybe something to keep in mind for future iterations.

Also: this could just be voodoo 😅

Reply:
Yeah, I go back and forth on this. For example, with `CodePoint` we call `CodePoint.unsafeFrom` internally a lot in a context where we _know_ the value is a code point, because we pulled it out of a `String`. In that case, I'd like the unsafe version to avoid the `Either` construction/destruction. OTOH, if the normative case is to call the safe version, then I tend to prefer having the unsafe version call the safe version.

In this case, I don't think it really matters since 99% of all users will be using `Damp.PunycodeDamp` :smile:

Reply:
> In that case, I'd like the unsafe version to avoid the Either construction/destruction

Not knowing anything, I suspect this can be inlined and the allocations can be vanished much more easily than optimizing a try catch.